File: CSharpCodeBase/components/character/aiworldmap.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class aiworldmap :component{
 public void init (self, entity);
       Component.init(self, entity);
       this._map = {}
       this.entity.worldmap = self;
 }})
 public void AddLocation(locName, vec3){
   this._map[locName] = vec3;
 }
 public void GetLocation(locName){
   return this._map[locName]  ||  this.entity:GetPosition();
 }
 public void GetDistanceTo(locName){
   return this.entity:GetSimpleDistanceToVec(self:GetLocation(locName));
 }
 public void Save(storage){
   var i = 0;
   for(k, v in pairs(this._map) ){
     storage:SetFloat("aiworldmap_" .. i .. "x", v.x);
     storage:SetFloat("aiworldmap_" .. i .. "y", v.y);
     storage:SetFloat("aiworldmap_" .. i .. "z", v.z);
     storage:SetString("aiworldmap_" .. i .. "key", k);
     i = i + 1;
   }
   storage:SetInt("aiworldmap_count", #this._map);
 }
 public void Load(storage){
   var count = storage:GetInt("aiworldmap_count", 0);
   if(count == 0  ){ 
     self:AddLocation("home", this.entity:GetPosition(true));
   }else{
     for(i = 0, count ){
       var x = storage:GetFloat("aiworldmap_" .. i .. "x");
       var y = storage:GetFloat("aiworldmap_" .. i .. "y");
       var z = storage:GetFloat("aiworldmap_" .. i .. "z");
       var key = storage:GetString("aiworldmap_" .. i .. "key") ;
       self:AddLocation(key, Vector3(x, y, z));
     }
   }
 }
 }}

File: CSharpCodeBase/components/character/aiworldmap.lua
local Class = require 'utils.hump.class'
local Component = require 'components.component'
local Vector3 = require 'utils.hump.vector3'

local AIWorldMap = Class({__includes=Component, init = function (self, entity)
      Component.init(self, entity)
      self._map = {}
      self.entity.worldmap = self
end})

function AIWorldMap:AddLocation(locName, vec3)
  self._map[locName] = vec3
end

function AIWorldMap:GetLocation(locName)
  return self._map[locName] or self.entity:GetPosition()
end

function AIWorldMap:GetDistanceTo(locName)
  return self.entity:GetSimpleDistanceToVec(self:GetLocation(locName))
end

function AIWorldMap:Save(storage)
  local i = 0
  for k, v in pairs(self._map) do
    storage:SetFloat('aiworldmap_' .. i .. 'x', v.x)
    storage:SetFloat('aiworldmap_' .. i .. 'y', v.y)
    storage:SetFloat('aiworldmap_' .. i .. 'z', v.z)
    storage:SetString('aiworldmap_' .. i .. 'key', k)
    i = i + 1
  end
  storage:SetInt('aiworldmap_count', #self._map)
end

function AIWorldMap:Load(storage)
  local count = storage:GetInt('aiworldmap_count', 0)
  if count == 0 then
    self:AddLocation('home', self.entity:GetPosition(true))
  else
    for i = 0, count do
      local x = storage:GetFloat('aiworldmap_' .. i .. 'x')
      local y = storage:GetFloat('aiworldmap_' .. i .. 'y')
      local z = storage:GetFloat('aiworldmap_' .. i .. 'z')
      local key = storage:GetString('aiworldmap_' .. i .. 'key') 
      self:AddLocation(key, Vector3(x, y, z))
    end
  end
end

return AIWorldMap